### DISCUSSION TITLE
`prisma generate` moves generated package outside of working directory

### DISCUSSION THREAD

**Discussion by ghost** [prisma/prisma#12724] [Q&A]
I also asked this question [here](https://stackoverflow.com/questions/71785435/prisma-generate-moves-generated-package-outside-of-working-directory) but figured that this might be a more appropriate place!

----------------

I have a setup where I want to share a Prisma schema file between different `node` packages within a mono-repo.
The schema has multiple but these models are not exported to the client. The reason is because the local client is not actually updated when running `prisma generate --schema=...`. It updates another instance of the client, as can be seen in the output of the command:

```text
$ npx prisma generate --schema=./../shared/prisma/prisma/schema.prisma
Prisma schema loaded from ../shared/prisma/prisma/schema.prisma

✔ Generated Prisma Client (3.12.0 | library) to ./../shared/prisma/node_modules/@prisma/client in 199ms
You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client

import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()

Done in 2.66s.
```

In the local package directly under `node_modules` the types are nonsensical:
```typescript
export declare const PrismaClient: any
...
export declare type PrismaClient = any
export declare const dmmf: any
export declare type dmmf = any
...
```

How do I make sure that the local copy of the client is updated? Is there something wrong in the way I approach this (is there a better way to share a Prisma schema between packages)?

Thanks in advance!

**Comment by janpio** [comment]
Can you share more, optimally a reproduction, of how your project is set up so we can reproduce and understand this?

**Reply by ghost** [reply]
Sure can!
[prisma-setup-main.zip](https://github.com/prisma/prisma/files/8463782/prisma-setup-main.zip)

In this ZIP you'll find an almost empty setup with a prisma schema in the `/schema` folder, and two packages; A and B. I want packages A and B to share the schema. I first tried running the following commands:
```
cd A
yarn
npx prisma generate --schema=./../prisma/schema.prisma
```

This seems to work fine, if I look into `A/index.ts` and inspect the PrismaClient typing, there seems to be a correct initialization of the types (I assume the underlying code is also correct).
If I then `cd` into B, and run `yarn`, followed by `npx prisma generate --schema=./../prisma/schema.prisma`, and inspect the `PrismaClient` in `B/index.ts`, the type resolves to `any`. In short; the typings generated from the same schema are not equal. The reason seems to be that `npx prisma generate ...` correctly links the generated package and type into A's `node_modules` the first time, but after that on subsequent `generates` from a different folder (B) the linking goes wrong.

The first time generating (A), the log prisma outputs indicate a correct linking to the packages `node_modules`:
```
✔ Generated Prisma Client (3.12.0 | library) to ./node_modules/@prisma/client in 79ms
```

When generating in package B this linking seems to be done to the `prisma` folder's node modules:
```
✔ Generated Prisma Client (3.12.0 | library) to ./../prisma/node_modules/@prisma/client in 87ms
```

Somehow it seems that the first `generate` creates a pseudo package inside the `prisma` folder but doesn't clean up, which misleades the second `generate`. I'm not too experienced with prisma to know the intricacies of the `generate` process but the `schema` seems to indicate that I can reference an out-of-folder schema without changing any behaviour.

**Reply by ghost** [reply]
I should note that I got it working now by keeping the schema in a share package instead of just in a folder. If I then import that "local" package after generating schemas, both A and B read the correct types. It just seems that the `--schema` flag does not seem to work as intended.

**Reply by ghost** [reply]
Diving a little bit deeper into using a shared package with a schema instead of just a shared schema;
It seems that the way the prisma client is being packaged throws off any importing from outside of the repo. I now have a setup where a package `prisma-schemas` is being imported by a node project using local imports; 
```
    "prisma-schemas": "../shared/prisma-schemas"
```
and when I want to rebuild I run something like this:
```
(cd ../shared/prisma-schemas && yarn build) && yarn add 'file:../shared/prisma-schemas'
```

However, on doing this the PrismaClient is installed locally in my node projects' `node_modules` but without the proper types. They default to `any` again.

**Reply by ghost** [reply]
Could this be a issue related to me using `yarn`? I retried the above setup with `npm` and then the modules are built as expected.

**Reply by janpio** [reply]
Yes, very much - but that is still something we want to get working.

I think you can turn your text here into a bug report issue (make sure to fill out the rest of the bug template as well please). Thanks.
